### ISSUE TITLE
Strange behavior when formatting macro inside closure

### ISSUE THREAD

**Issue by 20x48** [rust-lang/rustfmt#5170]
It's a picec of game code inside `event_loop` (provided by crate **glium**), a closure:

``` rust
            // framerate
            match cfg.video.framerate {
                video::Framerate::AtMost(n) => {
                    flag_framerate_limit = true;
                    time_interval_least = Duration::from_secs_f32(1. / n as f32);
                    trace!(g; "帧率限制已启用: {}FPS; 每帧最短停留时间: {}us", n, time_interval_least.as_micros());🎇
                }
                video::Framerate::Unlimited => {
                    flag_framerate_limit = false;
                    trace!(g; "帧率限制已禁用")
                },
            }
```

(Of course, many block and line comments inside there)

`🎇` tells you which line is the one I want pay attention to. Don't mind the Chinese here.

- Without making any changes to this line: *The entire closure cannot be formatted.*
  (But code outside the closure can still be formatted. By the way, this line has exactly 100 chars, same as the default. But this doesn't seem to be the point)
- If we delete all non-CJK chars in this line, it is now 85 chars: *can be just formatted now.*
- Or we delete `每帧最短停留时间`, make it 91 chars; Or delete "`每帧最短停留时` + any another non-CJK char", make it 92 chars: *can be just formatted now.*

From the above points, I can deduce that rustfmt is optimized for full-width chars. But this also doesn't seem to be the point.

- **Change `g;` to `g,`...** Wow! Now entire code can be formatted no matter how long this line is!

Here is the macro definitions I am using:

``` rust
macro_rules! __log {
    ($ch:expr; $lev:expr, $($det:tt)+) => (
        if $lev <= $ch.lev {
            $ch.send($lev, module_path!(), file!(), line!(), format!($($det)+))
        }
    )
}
macro_rules! fatal  { ($ch:expr; $($det:tt)+) => (__log!($ch; $crate::log::Level::Fatal,  $($det)+)) }
macro_rules! error  { ($ch:expr; $($det:tt)+) => (__log!($ch; $crate::log::Level::Error,  $($det)+)) }
macro_rules! warn   { ($ch:expr; $($det:tt)+) => (__log!($ch; $crate::log::Level::Warn,   $($det)+)) }
macro_rules! notice { ($ch:expr; $($det:tt)+) => (__log!($ch; $crate::log::Level::Notice, $($det)+)) }
macro_rules! info   { ($ch:expr; $($det:tt)+) => (__log!($ch; $crate::log::Level::Info,   $($det)+)) }
macro_rules! debug  { ($ch:expr; $($det:tt)+) => (__log!($ch; $crate::log::Level::Debug,  $($det)+)) }
macro_rules! trace  { ($ch:expr; $($det:tt)+) => (__log!($ch; $crate::log::Level::Trace,  $($det)+)) }
```

But this seems difficult to be reproduced in playground, perhaps because there is an extra nested macro.

Uh...that's all I can find. I'm not quite sure what information is useful. But this is bugging me, it's hard to modify a lot of macros scattered in many files, or writing code without formatting. By the way, I really don't want to change `g;` to `g,`, because log channel not belong to a message, they are different.

**Comment by 20x48** [comment]
I may understand why rustfmt refuses to wrap. If you replace this `;` with any other binary operator, for example, `+`, it will become:

``` rust
trace!(
    g + "message...",
    val1,
    val2,
);
```

So here is the comma version:

``` rust
trace!(
    g,
    "message...",
    val1,
    val2,
);
```

This problem is not actually related to CJK chars, it will also be the case when using letter padding long enough.

*Delimiters* similar to `;` are also `=>`, `->` and `]`, etc.

So...why not treat semicolons and commas as the same kind? And why if a macro cannot be formatted, the entire closure also cannot be formatted?

**Comment by calebcartwright** [comment]
Thanks for reaching out, but could I ask you to simplify the report a bit to make things more clear? Appreciate your attempts to help dig in, but it'd be cleaner and easier to have a succinct description of the issue with a complete/validly reproducible snippet, and we can separately discuss potential causes and workarounds if needed

**Comment by 20x48** [comment]
Ok, I will try to reproduce it clearly as soon as I am free.

**Comment by 20x48** [comment]
Reproduce it is easier than I thought. I believe it is enough as "Minimal... Example". Sorry for the delay of a week.

I have tried it on latest Rust: `rustfmt 1.4.38-stable (02072b48 2022-01-11)`

``` rust
// used to Send tracing to log channel...
macro_rules! trace { ($ch:ident; $($tail:tt)+) => (println!($($tail)+)) }
macro_rules! trace2 { ($ch:ident, $($tail:tt)+) => (println!($($tail)+)) }

fn main() {
    let _ch: ();

    // CASE 1: Closure in method call
    // ------------------------------
    //
    // * Here is a feature: "Maximum width" actually means
    //   "AFTER formatting, this line must be less than or equal to this width".
    // * Why "one hundred"? Because the default value of `max_width` in `rustfmt.toml`.

    Some(()).or_else(|| {
        // Like this, the line will actually increase an indentation level after formatting,
        // so its "maximum formatable width" is 96 chars instead of 100.
        //
        // Comments never count towards width.
        trace!(_ch; "A long long long message ................ (totally 97 width of this line)");
        let _able_to_format  :bool=false;         // Anything! Trailing space included!          
        None}).or_else(|| None).or_else(|| None); // What's worse? It's contagious!

    Some(())
        .or_else(|| {
            // After formatting, width of this line becomes exactly 100 chars instead of 96.
            trace!(_ch; "A long long long message ............... (exactly 96 width of this line)");
            let _able_to_format: bool = true;
            None
        })
        .or_else(|| None)
        .or_else(|| None);

    Some(())
        .or_else(|| {
            trace!(
                _ch;
                "Or I can make width of ANY LINE IN MACRO <= 100 chars manually ..................",
            );
            let _able_to_format: bool = true;
            None
        })
        .or_else(|| None)
        .or_else(|| None);

    // CASE 2: Independent closure
    // ---------------------------
    //
    // * No matter how, can always be formatted.

    (|| {
        // 103 chars width
        trace!(_ch; "abcdefg abcdefg abcdefg abcdefg abcdefg abcdefg abcdefg abcdefg abcdefg abcdefg");
        let _able_to_format: bool = true;
    })();

    // CASE 3: Use comma as delimiter instead of semicolon
    //         (closure also in method call)
    // ---------------------------------------------------
    //
    // * Hmm... WTF!?

    Some(()).or_else(|| {
        // It can be expanded automatically...
        //
        // - ONLY IF: After expansion, width of ANY LINE IN MACRO < 100 chars
        //            (Note that it is `<` here instead of `<=`)
        //            (Here is only one way to make it `==`: when this line ends with comma)
        // -    ELSE: Anything still can't be formatted!!!
        trace2!(_ch, "abcdefg abcdefg abcdefg abcdefg abcdefg abcdefg abcdefg abcdefg abcdefg abcdefg   ");
        let _able_to_format  :bool=false;
        None}).or_else(|| None).or_else(|| None);
}
```

**Comment by calebcartwright** [comment]
Thanks for following up, though going to close because it's a duplicate of #3863 and in part macro call related issues like #8.

tl;dr, rustfmt is a formatter of rust code, and args to macro calls are often not valid rust syntax. rustfmt will check to see if args can be formatted as valid rust syntax, and if so, will format them according to the corresponding rules. if not, then rustfmt largely leaves the macro calls alone/as the developer wrote them. As detailed in #3863, the intended behavior of rustfmt when it can't format something to fit within the value defined for max_width constraint, then it will bail and defer to the developer to format manually (that linked issue has some options you can turn on to get warnings/errors when this happens), and in the case of chains that means that if one chain item can't be formatted to fit then that cascades up to the entire chain.

So if you have something like a macro call that can't be formatted or a long string lit which exceed your max width boundary within a chain item, that produces the behavior you see where the chain isn't formatted

Also, not sure precisely what expansion was referring to in your last comment but note that rustfmt works exclusively in a pre-(macro) expansion context, because the output of rustfmt needs to be the same, valid, input code including for macro calls and their respective token streams
